test/infrastructure/container/interface.go

```diff
@@ -128,3 +131,26 @@ type Container struct {
 	// Status is the status of the container
 	Status string
 }
+
+// NewRuntimeClient gets a new instance of the right container runtime client
+// to use based on the current context.
+func NewRuntimeClient(ctx context.Context) (Runtime, error) {
+	if provider := RuntimeFromContext(ctx); provider != nil {
+		return provider, nil
+	}
+
+	// Anything else, including missing value, defaults to docker.
+	return NewDockerClient()
```

Comment:
My line of reasoning was that this testing change should not impact normal runtime operation. Since today CAPD will always communicate with docker, I thought we should keep that as a the default and treat any testing misses as a bug that would need to be fixed.

Long term, I think this call could be used to perform some detection and see whether we are running on a docker system, a podman system, or whatever other runtimes we may want to support in the future. But testing would need to be an explicit configuration that would override all else.

We could make specifying docker a requirement and fail otherwise. But this probably requires some thought into what that UX experience should be. How would this setting be propagated? There currently isn't a need to explicitly tell CAPD to use docker, so this would be adding a new configuration requirement to be able to specify that. Any thoughts on how best to handle that?

Reply:
I don't think this is are a requirement for testing, but it is a requirement for the entire "pluggable runtime" system.
Two use cases on top of my mind:
- https://github.com/kubernetes-sigs/cluster-api/issues/5634; this is CAPD using docker, but we want be able to use client with some specific configurations (connection to a remote docker server) and fall back to the default client will introduce unexpected behaviors/hide problems - some calls go to the remote server, some locally -
- https://github.com/kubernetes-sigs/cluster-api/issues/5146 podman or other runtimes

Last but not least, silent fallback could introduce unexpected behavior in tests as well.

If instead we remove defaulting, and we fix gaps in context propagation (from a quick check I don't expect big changes) we put ourself on a better path for the future.

Reply:
I wonder if Context should be used to pass this kind of information around.

Might be super bad idea but what about an alternative like storing the runtime "type" in a package-level unexported field and making it configurable with an exported func?

Reply:
I have seen using this approach extensively in knative or in the new Kubernetes test framework, so I'm ok with this as well as alternatives (even if I have a slight preference on Context).

My point is that whatever path we choose, we should always be explicit vs relying on "silent fallback" strategies that could introduce side effects

Reply:
Fine for me, just wanted to bring it up.

Reply:
Any thoughts on how best to configure the default docker context during normal runtime?

Reply:
What about setting it into the context in the main reconcile method?